server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/PlainHighlighter.java

```diff
@@ -157,10 +170,15 @@ public HighlightField highlight(FieldHighlightContext fieldContext) throws IOExc
             }
         }
 
-        // fragments are ordered by score by default since we add them in best
-        if (field.fieldOptions().scoreOrdered() == false) {
-            CollectionUtil.introSort(fragsList, (o1, o2) -> o1.getFragNum() - o2.getFragNum());
+        // For single text inputs, the fragments are already ordered by score. If we have multiple
+        // inputs, or if we are ordering by fragment number, then we need to resort the output list
+        if (textsToHighlight.size() > 1 || field.fieldOptions().scoreOrdered() == false) {
+            Comparator<OrderedTextFragment> comparator = field.fieldOptions().scoreOrdered()
+                ? Comparator.comparingDouble(OrderedTextFragment::score).reversed()
+                : Comparator.comparingInt(OrderedTextFragment::fragNum);
+            fragsList.sort(comparator);
```

Comment:
Just out of curiosity, did you consider making the list sorted from the start or would this have major drawbacks? I didn't clearly check whether sorting is always done here, maybe not and thats the reason why we don't need a sorted list from the start...

Reply:
The highlighter outputs fragments sorted by score, so if we only have a single input and we want score order then there's no need to do anything.  Otherwise we need to re-sort